tensorboard/plugins/pr_curve/tf_pr_curve_dashboard/tf-pr-curve-steps-selector.html

```diff
@@ -0,0 +1,194 @@
+      runToStep: {
+        type: Object,
+        notify: true,
+        computed: "_computeRunToStep(runToAvailableTimeEntries, _runToStepIndex)",
+      },
+      timeDisplayType: String,
+      // A mapping between run and slider index.
+      _runToStepIndex: Object,
+    },
+    listeners: {
+      "value-change": "_sliderValueChanged",
+      "immediate-value-change": "_sliderValueChanged",
+    },
+    observers: [
+      "_runsChanged(runs, _runToStepIndex)",
+      "_runToAvailableTimeEntriesChanged(runToAvailableTimeEntries)",
+    ],
+    ready() {
+      // Initialize the sliders so they point to the most recent value.
+      this._updateSliders(this._runToStepIndex);
+    },
+    _runsChanged(runs) {
+      // When selected runs change, make sure any new sliders point to the
+      // correct step. We use a setTimeout to add this operation to the end of
+      // the JS event queue so that it happens after the dom-repeat renders. We
+      // do not make this handler run when _runToStepIndex is updated because
+      // asynchronously changing _runToStepIndex could create a rendering loop
+      // and jittery sliders.
+      setTimeout(() => {
+        this._updateSliders(this._runToStepIndex);
+      }, 0);
+    },
+    _computeColorForRun(run) {
+      return runsColorScale(run);
+    },
+    _computeTimeTextForRun(
+        runToAvailableTimeEntries, runToStepIndex, run, timeDisplayType) {
+      const stepIndex = runToStepIndex[run];
+      if (!_.isNumber(stepIndex)) {
+        // The step is not known yet.
+        return '';
+      }
+
+      const entries = runToAvailableTimeEntries[run];
+      if (!entries) {
+        // No PR curve data has been received from the server yet.
+        return '';
+      }
+
+      const value = entries[stepIndex][timeDisplayType];
+      if (timeDisplayType === 'step') {
+        return `step ${value}`;
+      } else if (timeDisplayType === 'relative') {
+        // Return the time using the units that are most apt.
+        if (value < 1) {
+          return `${(value * 1000).toFixed(2)} ms`;
+        }
+        return `${value.toFixed(2)} s`;
+      } else if (timeDisplayType === 'wall_time') {
+        return new Date(value * 1000).toString();
+      }
+      console.error(
+          `The display type of ${timeDisplayType} is not recognized.`);
+    },
+    _sliderValueChanged(event) {
+      const run = event.target.dataset.run;
+      const newRunToStepIndex = _.clone(this._runToStepIndex);
+      newRunToStepIndex[run] = event.target.immediateValue;
+      this.set('_runToStepIndex', newRunToStepIndex);
+    },
+    _computeMaxStepIndexForRun(runToAvailableTimeEntries, run) {
+      const entries = runToAvailableTimeEntries[run];
+      return entries && entries.length ? entries.length - 1 : 0;
+    },
+    _runToAvailableTimeEntriesChanged(runToAvailableTimeEntries) {
+      // The step for each run just refreshed. Change every slider to point to
+      // the last available step.
```

Comment:
Wait, why? This behavior seems really strange. Suppose that an experiment is actively writing data to our logdir. Then, every thirty seconds, all the sliders will be shoved over to the end.

Would it suffice to make only newly added runs point to the last available step?

Reply:
+1, that previous behavior seems jarring, ie after a reload, the steps unexpectedly change. Now, we only set sliders for newly added runs to the last available step - we initialize sliders to point to the last step, but never override the step again.

One small thing to note is that sampling logic samples a constant N values, so the step will slightly differ across reloads, but I think that's much less jarring than overriding step to the last value.

Reply:
Agreed.